Allow WKWebpagePreferences to pass a list of content rule list identifiers to be disabled

whsieh opened the pull request:
#### d8024c41f86ab3fcdc746444cfd671d2699e4893
<pre>
Allow WKWebpagePreferences to pass a list of content rule list identifiers to be disabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=251645">https://bugs.webkit.org/show_bug.cgi?id=251645</a>
rdar://104453682

Reviewed by Alex Christensen.

Add a new property to provide a fine-grained way to disable `WKContentRuleList`s by identifier, in
the case when content blockers are not disabled using the extant SPI `-_contentBlockersEnabled`. See
below for more details.

Test: WebpagePreferences.DisableContentRuleListsByIdentifier

* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::actionsForResourceLoad const):

Add an argument to take a filter function that can be used to skip processing content rule lists
given the rule list identifier; use this to skip content rule lists that are explicitly disabled via
the new SPI.

(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForLoad):
* Source/WebCore/contentextensions/ContentExtensionsBackend.h:
(WebCore::ContentExtensions::ContentExtensionsBackend::actionsForResourceLoad):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::setDisabledContentExtensions):
(WebCore::DocumentLoader::disabledContentExtensions const):

Add a getter and setter for the new variant that determines whether (and which) content rule lists
we should disable.

* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::userContentExtensionsEnabled const): Deleted.
(WebCore::DocumentLoader::setUserContentExtensionsEnabled): Deleted.

Merge the existing bit for `m_userContentExtensionsEnabled` and the new set of disabled content
rule list identifiers into a single `std::variant`, which contains one of the two following
alternatives:

1. An enum flag indicating that content rule lists should be enabled or disabled globally.
2. A `HashSet` of content rule list identifiers to disable.

We also add a typedef, `DisabledContentExtensions`, to more cleanly encapsulate this data structure.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::reload):
* Source/WebCore/page/UserContentProvider.cpp:
(WebCore::ruleListFilter):
(WebCore::UserContentProvider::processContentRuleListsForLoad):
* Source/WebKit/Shared/WebsitePoliciesData.cpp:
(WebKit::WebsitePoliciesData::encode const):
(WebKit::WebsitePoliciesData::decode):
(WebKit::WebsitePoliciesData::applyToDocumentLoader):

Change the state here to also match that of `DocumentLoader`, by replacing the current boolean
`m_contentBlockersEnabled` flag with `DisabledContentExtensions` state instead, which determines
whether or not we want to globally enable, globally disable or selectively disable content
extensions for this navigation.

* Source/WebKit/Shared/WebsitePoliciesData.h:
* Source/WebKit/UIProcess/API/APIWebsitePolicies.cpp:
(API::WebsitePolicies::copy const):
(API::WebsitePolicies::contentBlockersEnabled const):
(API::WebsitePolicies::setContentBlockersEnabled):

Set `DisabledContentExtensions` to either `DisabledContentExtensionsMode::None` or
`DisabledContentExtensionsMode::All`.

(API::WebsitePolicies::setDisabledContentRuleListIdentifiers):

Set `DisabledContentExtensions` to the given set of rule list identifiers.

(API::WebsitePolicies::disabledContentRuleListIdentifiers const):
(API::WebsitePolicies::data):
* Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
(-[WKWebpagePreferences _setDisabledContentRuleListIdentifiers:]):
(-[WKWebpagePreferences _disabledContentRuleListIdentifiers]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferencesPrivate.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/load-image.html: Added.

Add an API test to exercise this new SPI.

Canonical link: <a href="https://commits.webkit.org/259930@main">https://commits.webkit.org/259930@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/25669c00f3a67cb99f0e4f6496fdfa06b877d150

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106442 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/15468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39260 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115629 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110349 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16957 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6693 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98642 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112208 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95852 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94752 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27496 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8700 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28848 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9232 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/5422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14854 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48394 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6862 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10781 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/contentextensions/ContentExtensionsBackend.h:
```diff
@@ -67,7 +67,8 @@ class ContentExtensionsBackend {
         Vector<DeserializedAction> actions;
     };
-    WEBCORE_EXPORT Vector<ActionsFromContentRuleList> actionsForResourceLoad(const ResourceLoadInfo&) const;
+    enum class ShouldSkipRuleList { No, Yes };
```
`enum class ShouldSkipRuleList : bool`?

whsieh replied:
Fixed!

achristensen07 commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -2535,6 +2535,17 @@ void DocumentLoader::contentFilterHandleProvisionalLoadFailure(const ResourceErr
+
+void DocumentLoader::setDisabledContentRuleListIdentifiers(const Vector<String>& identifiers)
+{
+    m_disabledContentRuleListIdentifiers.clear();
```
This should be m_disabledContentRuleListIdentifiers = WTFMove(identifiers)

whsieh replied:
So `m_disabledContentRuleListIdentifiers` is a `HashMap`, while `identifiers` is a `Vector` (I opted to make `identifiers` a vector in the policies/IPC plumbing, to avoid some unnecessary hashing when propagating that list).

Are you suggesting that I should make this a `HashMap` throughout the code?

achristensen07 commented on Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
```diff
@@ -172,8 +172,11 @@ auto ContentExtensionsBackend::actionsForResourceLoad(const ResourceLoadInfo& re
     const ResourceFlags flags = resourceLoadInfo.getResourceFlags() | ActionConditionMask;
-    for (auto& contentExtension : m_contentExtensions.values())
+    for (auto& [identifier, contentExtension] : m_contentExtensions) {
+        if (shouldSkipRuleList && shouldSkipRuleList(identifier) == ShouldSkipRuleList::Yes)
```
Don't do a null check inside the inner loop.  Make it always non-null.  You might even be able to use ScopedLambda

whsieh replied:
Good point! I'll refactor to avoid these extra null checks (here, as well as the `loader` check below).

Edit: tried to use `ScopedLambda`, but ran into a bit of trouble due to the fact that I need to capture a reference to the main document loader in the block. I think it would be cleaner (and safer) to just use a non-null `Function`.

achristensen07 commented on Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
```diff
@@ -216,8 +219,14 @@ ContentRuleListResults ContentExtensionsBackend::processContentRuleListsForLoad(
     ResourceLoadInfo resourceLoadInfo { url, mainDocumentURL, frameURL, resourceType, mainFrameContext };
-    auto actions = actionsForResourceLoad(resourceLoadInfo);
+    auto actions = actionsForResourceLoad(resourceLoadInfo, [loader = mainDocumentLoader](const String& identifier) {
+        return loader && loader->isDisabledContentRuleListIdentifier(identifier) ? ShouldSkipRuleList::Yes : ShouldSkipRuleList::No;
```
This is another unnecessary check inside the inner loop.  In fact, this check should be in the same place as what is currently in contentRuleListsEnabled.

achristensen07 commented on Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
```diff
@@ -152,6 +155,7 @@ class WebsitePolicies final : public API::ObjectImpl<API::Object::Type::WebsiteP
     // FIXME: replace most or all of these members with a WebsitePoliciesData.
     bool m_contentBlockersEnabled { true };
     HashMap<WTF::String, Vector<WTF::String>> m_activeContentRuleListActionPatterns;
+    Vector<WTF::String> m_disabledContentRuleListIdentifiers;
```
It would be helpful to prevent bad code design and forgetting checks and moves to combine this with m_contentBlockersEnabled and replace them both with an Expected<ContentBlockersEnabled, MemoryCompactRobinHoodHashSet<String>>

whsieh replied:
To clarify, do you mean the following?
1. Replace the `bool` flag with a named enum class, `ContentBlockersEnabled`.
2. Use a single `Expected<ContentBlockersEnabled, MemoryCompactRobinHoodHashSet>` in place of both `m_contentBlockersEnabled` and `m_disabledContentRuleListIdentifiers`

(As we discussed offline, I'll use a `std::variant` instead of an `WTF::Expected`)

achristensen07 commented on Source/WebCore/page/UserContentProvider.cpp:
```diff
@@ -93,16 +93,49 @@ void UserContentProvider::invalidateInjectedStyleSheetCacheInAllFramesInAllPages
     }
 
-    return true;
+    return std::visit(WTF::makeVisitor([](const HashSet<String>& disabledRuleLists) -> ContentExtensions::ContentExtensionsBackend::RuleListFilter {
```
I think this trailing return type is unnecessary.

whsieh replied:
Ah, I had tried to remove these, but got:

```
error: static_assert failed due to requirement '__all<false>::value' "`std::visit` requires the visitor to have a single return type."
```

In lieu of the trailing return types, I wrapped these in a `ContentExtensions::ContentExtensionsBackend::RuleListFilter { }`, which takes care of the above build error.

achristensen07 commented on Source/WebCore/page/UserContentProvider.cpp:
```diff
@@ -93,16 +93,49 @@ void UserContentProvider::invalidateInjectedStyleSheetCacheInAllFramesInAllPages
+{
+    RefPtr mainLoader = mainDocumentLoader(documentLoader);
+    if (!mainLoader)
+        return false;
```
This should be true to match previous behavior to make it fail safe.

whsieh replied:
Nice catch!

whsieh committed:
Allow WKWebpagePreferences to pass a list of content rule list identifiers to be disabled
https://bugs.webkit.org/show_bug.cgi?id=251645
rdar://104453682

Reviewed by Alex Christensen.

Add a new property to provide a fine-grained way to disable `WKContentRuleList`s by identifier, in
the case when content blockers are not disabled using the extant SPI `-_contentBlockersEnabled`. See
below for more details.

Test: WebpagePreferences.DisableContentRuleListsByIdentifier

* Source/WebCore/contentextensions/ContentExtensionsBackend.cpp:
(WebCore::ContentExtensions::ContentExtensionsBackend::actionsForResourceLoad const):

Add an argument to take a filter function that can be used to skip processing content rule lists
given the rule list identifier; use this to skip content rule lists that are explicitly disabled via
the new SPI.

(WebCore::ContentExtensions::ContentExtensionsBackend::processContentRuleListsForLoad):
* Source/WebCore/contentextensions/ContentExtensionsBackend.h:
(WebCore::ContentExtensions::ContentExtensionsBackend::actionsForResourceLoad):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::setDisabledContentExtensions):
(WebCore::DocumentLoader::disabledContentExtensions const):

Add a getter and setter for the new variant that determines whether (and which) content rule lists
we should disable.

* Source/WebCore/loader/DocumentLoader.h:
(WebCore::DocumentLoader::userContentExtensionsEnabled const): Deleted.
(WebCore::DocumentLoader::setUserContentExtensionsEnabled): Deleted.

Merge the existing bit for `m_userContentExtensionsEnabled` and the new set of disabled content
rule list identifiers into a single `std::variant`, which contains one of the two following
alternatives:

1. An enum flag indicating that content rule lists should be enabled or disabled globally.
2. A `HashSet` of content rule list identifiers to disable.

We also add a typedef, `DisabledContentExtensions`, to more cleanly encapsulate this data structure.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::reload):
* Source/WebCore/page/UserContentProvider.cpp:
(WebCore::ruleListFilter):
(WebCore::UserContentProvider::processContentRuleListsForLoad):
* Source/WebKit/Shared/WebsitePoliciesData.cpp:
(WebKit::WebsitePoliciesData::encode const):
(WebKit::WebsitePoliciesData::decode):
(WebKit::WebsitePoliciesData::applyToDocumentLoader):

Change the state here to also match that of `DocumentLoader`, by replacing the current boolean
`m_contentBlockersEnabled` flag with `DisabledContentExtensions` state instead, which determines
whether or not we want to globally enable, globally disable or selectively disable content
extensions for this navigation.

* Source/WebKit/Shared/WebsitePoliciesData.h:
* Source/WebKit/UIProcess/API/APIWebsitePolicies.cpp:
(API::WebsitePolicies::copy const):
(API::WebsitePolicies::contentBlockersEnabled const):
(API::WebsitePolicies::setContentBlockersEnabled):

Set `DisabledContentExtensions` to either `DisabledContentExtensionsMode::None` or
`DisabledContentExtensionsMode::All`.

(API::WebsitePolicies::setDisabledContentRuleListIdentifiers):

Set `DisabledContentExtensions` to the given set of rule list identifiers.

(API::WebsitePolicies::disabledContentRuleListIdentifiers const):
(API::WebsitePolicies::data):
* Source/WebKit/UIProcess/API/APIWebsitePolicies.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferences.mm:
(-[WKWebpagePreferences _setDisabledContentRuleListIdentifiers:]):
(-[WKWebpagePreferences _disabledContentRuleListIdentifiers]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebpagePreferencesPrivate.h:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/load-image.html: Added.

Add an API test to exercise this new SPI.

Canonical link: https://commits.webkit.org/259930@main
